go/vt/vtgate/semantics/transitive_closures.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2025 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package semantics
+
+import (
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+type (
+	// TransitiveClosures expression equality, so we can rewrite expressions efficiently
+	// It allows for transitive closures (e.g., if a == b and b == c, then a == c).
+	TransitiveClosures struct {
+		// ExprIDs stores the IDs of the expressions.
+		// The ID can then be used to get other equivalent expressions from the Equalities slice
+		exprIDs map[sqlparser.Expr]int
+
+		// notMapKeys stores expressions that are not valid as map keys
+		notMapKeys []ExprWithID
+
+		// Both the above fields are used to find the offset into this slice
+		equalities [][]sqlparser.Expr
+
+		comparer func(a, b sqlparser.Expr) bool
+	}
+
+	// ExprWithID is used to keep track of expressions that are not valid as map keys
+	ExprWithID struct {
+		Expr sqlparser.Expr
+		ID   int
+	}
+
+	SemanticColName struct {
+		Name      sqlparser.IdentifierCI
+		Qualifier sqlparser.TableName
+	}
+)
+
+// NewTransitiveClosures creates a new TransitiveClosures
+func NewTransitiveClosures() *TransitiveClosures {
+	return &TransitiveClosures{
+		exprIDs: make(map[sqlparser.Expr]int),
+	}
+}
+
+func (ce *TransitiveClosures) getID(e sqlparser.Expr, comparer func(a, b sqlparser.Expr) bool) (int, bool) {
+	if !ValidAsMapKey(e) {
+		for _, id := range ce.notMapKeys {
+			if comparer(id.Expr, e) {
+				return id.ID, true
+			}
+		}
+
+		return 0, false
+	}
+
+	// first let's try to find the expression in the map
+	id, found := ce.exprIDs[e]
+	if found {
+		return id, found
+	}
+
+	// we might still have a match in the map, just not by reference
+	for expr, id := range ce.exprIDs {
+		if comparer(expr, e) {
+			defer func() {
+				// we do this in the defer so we don't mutate the map while iterating over it
+				ce.exprIDs[e] = id
+			}()
+			return id, true
+		}
+	}
+
+	return 0, false
+}
+
+func (ce *TransitiveClosures) setID(in sqlparser.Expr, id int, comparer func(a, b sqlparser.Expr) bool) {
```

Comment:
add a comment that the `equalities` slice has to be updated before calling this method, like we do in `Add`

Reply:
da267b069bd